### PR TITLE
Warn on receiving a space before the token

### LINKED ISSUE #106142: kube-apiserver incorrectly parses tokens starting with ' '
### What happened?

kube-apiserver authenticated the token " TOKEN" as `system:anonymous` (note the leading space).

### What did you expect to happen?

kube-apiserver should reject the invalid authorization header.

The bearer token [RFC](https://datatracker.ietf.org/doc/html/rfc6750#section-2.1) specifies a single space between "Bearer" and the token value.

### How can we reproduce it (as minimally and precisely as possible)?

```
$ IP=<address of kube-apiserver>
$ curl -k https://IP -H "Authorization: Bearer asdf"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}
$ curl -k https://IP -H "Authorization: Bearer  asdf"
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "forbidden: User \"system:anonymous\" cannot get path \"/\"",
  "reason": "Forbidden",
  "details": {
    
  },
  "code": 403
}
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.18-dispatcher", GitCommit:"de944d802c49735ad0f4bfe82ddfa19737ebe962", GitTreeState:"clean", BuildDate:"2021-05-13T17:51:55Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.20-gke.901", GitCommit:"db8463aecdfe0eb5d9067effa2fd5ab3ff7a988e", GitTreeState:"clean", BuildDate:"2021-08-04T23:27:42Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by slaskawi**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This Pull Request adds a warning when someone tries to send a request with a space before the token:

```
$ curl -k https://IP -H "Authorization: Bearer  asdf"
```

#### Which issue(s) this PR fixes:

xref #106142

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by slaskawi** [comment]
@aramase Could I ask you for a review?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go [hunk 1]
```diff
@@ -48,6 +50,10 @@ func (a *Authenticator) AuthenticateRequest(req *http.Request) (*authenticator.R
 
 	// Empty bearer tokens aren't valid
 	if len(token) == 0 {
+		// The space before the token case
+		if len(parts) == 3 {
+			klog.Warning("WARNING: the support for spaces before the token will be removed shortly")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:55]: A kube-apiserver log warning is not visible to the person making the request, who is the one that is likely to be confused by lack of authentication and is the one who needs to make a change. Also, we haven't decided that this type of request will turn into an error in the future, so don't include that in the message.

This should be something like this:
```go
warning.AddWarning(req.Context(), "", "the provided Authorization header contains extra space before the bearer token, and is ignored")
```

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken_test.go [hunk 2]
```diff
@@ -184,6 +184,13 @@ func TestBearerToken(t *testing.T) {
 			ExpectedErr:                  true,
 			ExpectedAuthorizationHeaders: []string{"Bearer 123"},
 		},
+		"valid bearer token with a space": {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken_test.go:187]: you can add a dummy warning recorder to the context in the dummy request passed into the unit test (see dummyRecorder in pkg/registry/batch/job/storage/storage_test.go for an example) and verify the warning was added here

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go (hunk 1 above)
Comment (slaskawi) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:55]: Sure! I just changed the text slightly: `the provided Authorization header contains extra space before the bearer token, which is ignored`. I hope it's OK.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken_test.go (hunk 2 above)
Comment (slaskawi) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken_test.go:187]: Done!

**Comment by slaskawi** [comment]
Thanks a lot for the review @liggitt !

I addressed all the comments according to your guidance. Please have another look once CI is happy.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go [hunk 3]
```diff
@@ -22,6 +22,11 @@ import (
 	"strings"
 
 	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/warning"
+)
+
+const (
+	invalidTokenWithSpaceWarning = "the provided Authorization header contains extra space before the bearer token, which is ignored"
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/bearertoken/bearertoken.go:29]: this is hard to tell whether the extra space is ignored, or the bearer token is ignored

**Comment by slaskawi** [comment]
The `[sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support non-existent path expand_more` failure doesn't seem to be related. Let's retrigger it.

/retest

**Comment by slaskawi** [comment]
@enj @liggitt Could you please advice me what milestone we're targeting at with this PR? `tide` is complaining that `v1.24` is not set.

**Pull request merged**